### PR TITLE
feat!: expose raw transport responses

### PR DESCRIPTION
**Pull request by bitpshr**
This PR updates all endpoints to expose full responses using a new API.

```js
const branches = await client.branches.list({ ... });
const response = client.unwrap(branches);
```

In the example above, `response` would be the un-sanitized JSON response containing data exactly as it's returned from a given transport. The advantage of using a helper function per @amccloud's suggestion is that consumers aren't forced to grab an arbitrary, potentially-changing property on either the returned object or the returned array, the latter of which felt especially odd.

Internally, we're able to hide the added response key entirely using a non-enumerable object property.

References #192
References #193 
References #190

### CODE REVIEW

File: src/endpoints/Branches.js [hunk 1]
```diff
@@ -7,23 +7,28 @@ import type {
...
   info(descriptor: BranchDescriptor, requestOptions: RequestOptions = {}) {
     return this.configureRequest<Promise<Branch>>({
-      api: () => {
-        return this.apiRequest(
+      api: async () => {
+        const response = await this.apiRequest(
           `projects/${descriptor.projectId}/branches/${descriptor.branchId}`
         );
+
+        return wrap(response, response);
```
Comment (bitpshr) [inline src/endpoints/Branches.js]: `branches.info` is a weird method to look at since the `response` is the actual `value`. This was only piped . through the `wrap` flow for consistency. `branches.list` is a better example to check out.

File: src/response.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// @flow
+export function wrap(value: any, response: any) {
```
Comment (bitpshr) [inline src/response.js]: All uses of `wrap` still end up returning strongly-typed values because of upstream generics. We could definitely add a generic here for readability, but it'd force every consumer to redefine the return value type unnecessarily.

File: src/endpoints/Branches.js (hunk 1 above)
Comment (tommoor) [reply src/endpoints/Branches.js]: Hah, I guess in the case that they are identical we could just pass the single argument and have `wrap` handle that?

File: src/response.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  value = Array.isArray(value) ? [...value] : { ...response };
```
Comment (amccloud) [inline src/response.js]: Should this be `{...value}`?

also do we need to clone the value or can we define `_response` on the original value object?

File: src/response.js (hunk 2 above)
Comment (amccloud) [reply src/response.js]: Upstream values are `any` so it might not be a poroblem. A generic would allow you to type this locally. We know `value` must be object like for us to define a property on it and `_response` can be `mixed` or also just an object

```js
type Value = {} | [];
type Envelope<T: Value> = T & { _response?: {} };

export function wrap<T: Value>(value: Envelope<T>, response: {}): Envelope<T> {
  Object.defineProperty(value, "_response", {
    enumerable: false,
    value: response
  });

  return value;
}

export function unwrap<T: {}>(value: Envelope<T>) {
  return value._response;
}
```

File: src/response.js (hunk 3 above)
Comment (bitpshr) [reply src/response.js]: Great catch re: `{ ...value }`, fixed.

We need to clone the value to prevent a circular reference (adding `_response` to `value` also modifies that data in-place on the original `response`.)

Comment (bitpshr) [reply src/response.js]: Thinking about this more, we can clone `response` instead, which simplifies this considerably (as well as the typings.)

File: src/response.js (hunk 2 above)
Comment (bitpshr) [reply src/response.js]: Good suggestion, updated.

File: src/endpoints/Branches.js (hunk 1 above)
Comment (bitpshr) [reply src/endpoints/Branches.js]: Updated.

File: docs/abstract-api.md [hunk 4]
```diff
@@ -246,16 +246,6 @@ abstract.branches.info({
...
-### The ChangesetResponse object
-
-| Property | Type | Description |
-|----------|------|-------------|
-| `changeset` | `Changeset` | Base changeset |
-| `files` | `File[]` | Files that changed in this changeset |
-| `layers` | `Layer[]` | Layers that changed in this changeset |
-| `pages` | `Page[]` | Pages that changed in this changeset |
-| `sharedData` | `SharedItem[]` | Shared data that changed in this changeset |
-
```
Comment (bitpshr) [inline docs/abstract-api.md]: We can stop maintaining the few `<entity_type>Response` types we had defined now that this data will be exposed via `client.unwrap(entity)`. Admittedly, it does feel weird that we no longer have documentation for these related entity types though. If we keep maintaining the documentation for these response pages (e.g. still maintaining a new concept for each response), why not just maintain and return the type for build-time safety, too?

TLDR; We need to figure out if and how we plan to document this response meta now that the types went away.

**Comment by bitpshr** [comment]
Ready for a pass when you have a chance @tommoor and @amccloud.

File: docs/abstract-api.md (hunk 4 above)
Comment (tommoor) [reply docs/abstract-api.md]: > Admittedly, it does feel weird that we no longer have documentation for these related entity types though

Yep, well I guess this was only supposed to be a stepping stone to move us forward. I could see us having documentation for our actual API also which would help here – but ultimately any objects which can only be accessed through unwrap need their own endpoints, e.g libraries.

Comment (bitpshr) [reply docs/abstract-api.md]: Even with separate endpoints for every entity type, we'd still need a way to tell consumers what related entity types are wrapped up in a given response. For example, even though we have a separate endpoint for `files` already, there's no indication that `files` even exists in the wrapped response for `changesets`. Internal caching may help, but I'm still a little skeptical of a cache-based approach (e.g. to get pages related to changesets, a consumer would have to fetch all changesets, then use the same `BranchDescriptor` to fetch all files, then iterate over each file and request each page, etc.)

I agree that we're getting to the point where we need API documentation to better support this need and also external contributions.

**Comment by tommoor** [comment]
Wanna publish a prerelease and I'll try and integrate internally?

**Comment by bitpshr** [comment]
Definitely @tommoor. `abstract-sdk@8.0.0-beta.3` is now live for testing.
